README.md
# ОФД

Библиотека для работы с бинарным протоколом, используемым кассовыми аппаратами для обмена данными с оператором фискальных данных (ОФД).
Подробнее об ОФД и новом порядке применения кассовой технике можно прочитать здесь: <https://kkt-online.nalog.ru>

ФФД (формат фискальных данных) - бинарный формат, используемый для обмена сообщениями между кассой и ОФД.
Для передачи фискальных документов от ОФД в налоговую (ФНС) используется описание документов в JSON формате.

## Распаковка сообщения от кассы

```python
import ofd

message = b'\x5a\x01\x5d\xa8\xa5...'  # Контейнер сообщения от кассы в бинарном виде.
fiscal_sign = b'\x23\14\12..'  # Фискальный признак документа в бинарном формате - дописывается в конец поля rawData.
doc = ofd.unpack_container_message(message, fiscal_sign)
```

## Упаковка json документа в бинарный формат

```python
import ofd

doc = {'receipt': { .. }}  # Фискальный документ в JSON формате.
message = ofd.pack_json(doc, ofd.DOCS_BY_NAME)  # Получаем контейнер ФФД в бинарном формате.
```

## Запуск тестов

```bash
python3.5 setup.py pytest
```

## Эмулятор ОФД

В директории ./example написан эмулятор ОФД, демонстрирующий использование протокола. Это TCP-сервер, которое слушает заданный порт.
Если отправить на вход данные бинарного протокола, то приложение расшифрует сообщение в json-формат и выведет его в stdout.
В ответ эмулятор отправит документ "подтверждение оператора" в бинарном формате. Эмулятор работает без использования
шифровальной машины, поэтому сможет прочитать только незашифрованные входящие сообщения. Пример отправки документа можно посмотреть в tests/example_test.py
Запустить пример можно следующей командой, указав желаемый номер порта:

```bash
python3.5 example/mock_ofd.py --port 12345
```

Для запуска под windows

```
py -3.5 example/mock_ofd.py --port 12345
```


CHANGELOG.md
# Changelog

## Unreleased (2022-11-03)

#### New Features

* merge dev config in one, add utils ([#64](https://github.com/Mcublog/ofd/issues/64))
* debianization, JWT generation
* implement protocol from 19.10.2016
* implement 1101 tag
* complete validation part
* strict validation, part 1
* update JSON schema
* update protocol
* schema for close shift
* further scheme improvements
* improve scheme on items
* JSON schema for receipt
* use JSON schema for documents
* proper tracing
* crypto module can now be configured
* temporary suppress protocol check for new KKTs
* do not ignore cardinality
* multiple changes
* implement KGB cipher
* update YT schema
* update protocol to ATOL-3
* further improvements
* append into YT instead of rewrite
* add all possible tags
* add session header packer
* implemented responding
* working version
* initial commit
#### Fixes

* add parents to tags, bump 0.24.4
* allow non-unique correction code, bump 0.24.3
* remove user from required in openShift. bump
* protocol fix ([#12](https://github.com/Mcublog/ofd/issues/12))
* fixed correction receipt types in schema
* fix schemas
* use proper taxation_type tag for reports
* protocol compatibility with v2
* add missing protocol field
* add more protocol tags
#### Refactorings

* tests, add CI
* fix validators to match new protocol
* types
* new data
* dead code elimination
* protocol switch
* fiscalSign is a number now
* accept both 1 and 2 A-protocol
* proper cardinality handling
* make some methods static
* early corrupted protocol detection
* server decomposition, add unit tests
* adapt with python 3
* decomposition
* drop another magic constant
* calculate CRC for frame header
* drop magic constant
* frame headers now has its own class
* add stub for CRC
* add more tests
#### Docs

* fix python in examples, some wording


tests/example_test.py
import asyncio
import contextlib
import socket

import pytest

from example.mock_ofd import handle_connection, unpack_incoming_message


def unused_tcp_port() -> int:
    """Find an unused localhost port from 1024-65535 and return it."""
    with contextlib.closing(socket.socket(type=socket.SOCK_STREAM)) as sock:
        sock.bind(("127.0.0.1", 0))
        return sock.getsockname()[1]

@pytest.mark.asyncio(True)
async def test_ofd_emulation():
    binary_dump = b'*\x08A\n\x81\xa2\x00\x019999078900005488\xb4\x01\x14\x00\x00\x00\xb4\x01%x\xa5\x0b\x01\x10\t\x99\x99' \
               b'\x07\x89\x00\x00T\x88\x00\x00\x01\x84\xecL\x14\xc2\x00\x00\x01\x00\x04\x01\x8a\x0b\x00\x86\x01\x11' \
               b'\x04\x10\x009999078900005488\r\x04\x14\x000000000005008570    \xfa\x03\x0c\x007702203276  \x10\x04' \
               b'\x04\x00\x01\x00\x00\x00\xf4\x03\x04\x00\x98U\xb9X5\x04\x06\x00!\x04\xaa\x10uA \x04\x01\x00\x00' \
               b'\xea\x03\x01\x00\x00\xe9\x03\x01\x00\x00U\x04\x01\x00\x01V\x04\x01\x00\x00T\x04\x01\x00\x00&\x04' \
               b'\x01\x00\x06M\x04\x01\x00\x01\xf5\x03\x0c\x00000000000002\x18\x04\x11\x00\x8e\x8e\x8e \x90\x80\x8f' \
               b'\x8a\x80\x92-\xe6\xa5\xad\xe2\xe0 \xf1\x030\x00111141 \xa3.\x8c\xae\xe1\xaa\xa2\xa0, \xe3\xab. \x8a' \
               b'\xe3\xe1\xaa\xae\xa2\xe1\xaa\xa0\xef \xa4.20\x80 \xae\xe4\xa8\xe1 \x82-202\xf9\x03\x0c' \
               b'\x007704358518  $\x04\x08\x00nalog.ru]\x04\x13\x00example@example.com\xfd\x03\n\x00\x98\xa5\xad\xad' \
               b'\xae\xad \x8a. \x16\x04\n\x00OOO TAXCOM\xa5\x04\x01\x00\x02\xb9\x04\x01\x00\x02\xa4\x04\x03\x002.0' \
               b'\xa3\x041\x00111141 \xa3.\x8c\xae\xe1\xaa\xa2\xa0, \xe3\xab. \x8a\xe3\xe1\xaa\xae\xa2\xe1\xaa\xa0' \
               b'\xef\r\n\xa4.20\x80 \xae\xe4\xa8\xe1 \x82-202\xb3\x04\x0c\x00771234567890\xc5\x04\x01\x00\x00\xb7' \
               b'\x04\x01\x00\x01\x81\x06\xa5Z\xe1\x0cMu\x00\x00'

    port = unused_tcp_port()
    print(port)
    server = await asyncio.start_server(handle_connection, port=port)
    rd, wr = await asyncio.open_connection(port=port)

    wr.write(binary_dump)
    await wr.drain()
    try:
        doc, session, header = await unpack_incoming_message(rd)
        assert 'operatorAck' in doc
        assert doc['operatorAck']['fiscalDriveNumber'] == '9999078900005488'
        assert doc['operatorAck']['fiscalDocumentNumber'] == 1
        assert doc['operatorAck']['ofdInn'] == '7704358518'
        assert doc['operatorAck']['messageToFn'] == {'ofdResponseCode': 0}
    finally:
        server.close()


example/mock_ofd.py
# coding: utf8
#
#        Copyright (C) 2017 Yandex LLC
#        http://yandex.com
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#

import argparse
import asyncio
import json
import time

from ofd.protocol import (DOCS_BY_NAME, DocCodes, FrameHeader, SessionHeader,
                          String, pack_json, unpack_container_message)


async def unpack_incoming_message(rd):
    """
    Прочитать входящий поток бинарных данных и распаковать их в json документ
    """
    session_raw = await rd.readexactly(SessionHeader.STRUCT.size)
    session = SessionHeader.unpack_from(session_raw)
    print(session)
    container_raw = await rd.readexactly(session.length)
    header_raw, message_raw = container_raw[:FrameHeader.STRUCT.size], container_raw[FrameHeader.STRUCT.size:]
    header = FrameHeader.unpack_from(header_raw)
    print(header)
    return unpack_container_message(message_raw, b'0')[0], session, header


def create_response(doc, in_session, in_header):
    """
    Запаковать в протокол "подтверждение оператора" от ОФД к кассе
    :param doc: полученный документ
    :param in_header: заголовок контейнера входящего сообщения
    :param in_session: заголовок сессии входящего сообщения
    :return:
    """
    doc_body = doc[next(iter(doc))]  # получаем тело документа
    message = {
        'operatorAck': {
            'ofdInn': '7704358518',  # ИНН Яндекс.ОФД
            'fiscalDriveNumber': doc_body.get('fiscalDriveNumber'),
            'fiscalDocumentNumber': doc_body.get('fiscalDocumentNumber'),
            'dateTime': int(time.time()),
            'messageToFn': {'ofdResponseCode': 0}  # код ответа 0 при успешном получении документа
            # Теги ФПО и ФПП не указаны, т.к. должны быть добавлены реальным шифровальным комплексом
        }
    }
    message_raw = pack_json(message, docs=DOCS_BY_NAME)

    # в реальных ОФД FrameHeader формируется автоматически шифровальной машиной
    out_header = FrameHeader(length=FrameHeader.STRUCT.size + len(message_raw),
                             crc=0,
                             doctype=DocCodes.OPERATOR_ACK,
                             devnum=in_header.devnum,
                             docnum=String.pack(str(doc_body.get('fiscalDocumentNumber'))),
                             extra1=in_header.extra1,
                             extra2=String.pack('0'.rjust(12)))

    out_header.recalculate_crc(message_raw)
    container_raw = out_header.pack() + message_raw

    out_session = SessionHeader(pva=in_session.pva, fs_id=in_session.fs_id, length=len(container_raw), crc=0,
                                flags=0b0000000000010100)

    return out_session.pack() + container_raw


async def handle_connection(rd, wr):
    """
    Пример использования протокола для эмуляции работы ОФД. Сервер принимает входящее сообщение и распаковывает его,
    выводя значения в stdout. В ответ сервер формирует сообщение "подтверждение оператора" и передает его обратно кассе.
    Эмулятор работает без использования шифровальный машины, поэтому считаем, что сообщение приходит в ОФД
    в незашифрованном виде.
    :param rd: readable stream.
    :param wr: writable stream.
    """
    try:
        doc, session, header = await unpack_incoming_message(rd)
        print(json.dumps(doc, ensure_ascii=False, indent=4))
        response = create_response(doc, in_session=session, in_header=header)
        print('raw response', response)
        wr.write(response)
    finally:
        wr.write_eof()
        await wr.drain()


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--host', default=None, help='хост для запуска сервера')
    parser.add_argument('--port', default=12345, type=int, help='порт для запуска сервера')
    argv = parser.parse_args()
    host = None if argv.host in ['::', 'localhost'] else argv.host

    loop = asyncio.get_event_loop()
    server = asyncio.start_server(handle_connection, host=host, port=argv.port)
    loop.run_until_complete(server)
    print('mock ofd server has been started at port', argv.port)

    try:
        loop.run_forever()
    except KeyboardInterrupt:
        print('received SIGINT, shutting down')

    server.close()

    loop.run_until_complete(server.wait_closed())
    loop.close()
